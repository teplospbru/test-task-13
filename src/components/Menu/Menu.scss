@import '../../scss/variables';
@import '../../scss/fonts';

.menu {
    border-top: 1px solid $light-color;
    @include flex(space-between, center);
    padding: 1em 31px;

    @media  screen and (max-width: 767px) {
        margin: 1.5em 0;
        flex-wrap: wrap;
        //justify-content: center;
        gap:1.5em;
    }
    @media  screen and (max-width: 600px) {
        justify-content: center;
    }
    span {
        text-decoration: none;
        color: $primary-color;
        @include font(16, 400, 1);
        cursor: pointer;
    }
    > div {
        @include flex(space, center);
        gap: 0.5em;
        span {
            display: block;
            padding: 0.5em 1em;
            border: 2px solid #fff;
            border-radius: 6px;
            transition: 0.2s all;
            
            &.active, &:hover {
                border: 2px solid $red-color;
            }
        }
    }
}