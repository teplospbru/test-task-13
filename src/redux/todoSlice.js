import { createSlice } from "@reduxjs/toolkit";

export const todoSlice = createSlice({
    name: 'todos',
    initialState: {
        todos: [
            { id: 1, body: 'Next, we need to import the reducer function from the counter slice and add it to our store', isCompleted: false },
            { id: 2, body: 'Creating a slice requires a string name to identify the slice', isCompleted: true },
            { id: 3, body: 'Provide the Redux Store to React', isCompleted: false },
        ],
    },
    reducers: {
        createTodo(state, action) {
            state.todos.push({ id: Date.now(), body: action.payload, isCompleted: false })
        },
        setCompleted(state, action) {
            const index = state.todos.findIndex(todo => todo.id == action.payload);
            state.todos[index].isCompleted = !state.todos[index].isCompleted;
        },
        clearCompleted(state, action) {
            state.todos = state.todos.filter(todo => todo.isCompleted === false);
        }
    }
});

export const { clearCompleted, createTodo, setCompleted } = todoSlice.actions;

export default todoSlice.reducer;