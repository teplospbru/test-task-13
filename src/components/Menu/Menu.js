import { useDispatch } from 'react-redux';
import { clearCompleted } from '../../redux/todoSlice';
import './Menu.scss';

const Menu = ({ isFilteredBy, setFilteredBy, todos }) => {
    const dispatch = useDispatch();

    const uncompleted = todos.filter(todo => todo.isCompleted === false);

    const clearCompletedHandler = () => {
        dispatch(clearCompleted());
    }

    return (
        <nav className='menu'>
            <span>{ uncompleted.length + ' items left' }</span>
            <div>
                <span className={ isFilteredBy === 'All' ? 'active' : null } onClick={ () => setFilteredBy('All') }>All</span>
                <span className={ isFilteredBy === 'Active' ? 'active' : null } onClick={ () => setFilteredBy('Active') }>Active</span>
                <span className={ isFilteredBy === 'Completed' ? 'active' : null } onClick={ () => setFilteredBy('Completed') }>Completed</span>
            </div>
            <span onClick={ clearCompletedHandler }>Clear completed</span>
        </nav>
    )
};

export default Menu;