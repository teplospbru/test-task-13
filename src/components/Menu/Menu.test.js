import Menu from './Menu';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';
import renderer from 'react-test-renderer';

const todos = [
    { id: 1, body: 'Next, we need to import the reducer function from the counter slice and add it to our store', isCompleted: false },
    { id: 2, body: 'Creating a slice requires a string name to identify the slice', isCompleted: true },
    { id: 3, body: 'Provide the Redux Store to React', isCompleted: false },
];

const isFilteredBy = 'All';

beforeEach(() => {
    render(
        <Provider store={ store }>
            <Menu todos={ todos } isFilteredBy={ isFilteredBy } />
        </Provider>
    )
})

describe('testing the component "Menu"', () => {
    it('renders "Menu"', () => {
        const menu = screen.getByRole('navigation');

        expect(menu).toBeInTheDocument();
    });
    ['2 items left', 'All', 'Active', 'Completed', 'Clear completed'].forEach(item => {
        it(`renders menu item "${item}"`, () => {
            const text = screen.getByText(item);

            expect(text).toBeInTheDocument();
        })
    });
    it(`renders border around menu item "All"`, () => {
        const text = screen.getByText('All');

        expect(text).toHaveClass('active');
    });
    it('matches to snapshot', () => {
        const container = renderer.create(<Provider store={ store }><Menu todos={ todos } isFilteredBy={ isFilteredBy } /></Provider>).toJSON();

        expect(container).toMatchSnapshot()
    });
})