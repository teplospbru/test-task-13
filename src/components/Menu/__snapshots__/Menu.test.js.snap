// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the component "Menu" matches to snapshot 1`] = `
<nav
  className="menu"
>
  <span>
    2 items left
  </span>
  <div>
    <span
      className="active"
      onClick={[Function]}
    >
      All
    </span>
    <span
      className={null}
      onClick={[Function]}
    >
      Active
    </span>
    <span
      className={null}
      onClick={[Function]}
    >
      Completed
    </span>
  </div>
  <span
    onClick={[Function]}
  >
    Clear completed
  </span>
</nav>
`;
